### ISSUE TITLE
BadImageFormatException in Prime unit test when --optimize is disabled.

### ISSUE THREAD

**Issue by bryanedds** [Microsoft/visualfsharp#1257]
#### Repro steps
1. Pull down Prime code library from https://github.com/bryanedds/Prime/commit/43ad2039c51232c554339f820cebbcef794d7120
2. Open the Prime.sln solution found in ./Prime/Prime folder (should open with VS2015).
3. Ensure program configuration is set to Debug (should be by default).
4. Hit F5 to run program.
#### Expected behavior

The program executes and passes a single unit test called EventTests.productWorks () in Program.fs, then exits with code 0.
#### Actual behavior

Program halts with a BadImageFormatException in Observation.fs with error -

```
Exception thrown: 'System.BadImageFormatException' in Prime.exe

Additional information: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
```
#### Known workarounds

try loosening up the one or more of the type arguments with the _ wildcard at and around where the exception appears. If done correctly, it should still infer a workable secondary, if superfluous, parameter type. If that doesn't work, enable --optimize+ in F# project settings. The latter is not a good workaround since it cripples debugging tools like variable inspection and granular breakpoints.
#### Related information

I ran the generated executable through peverify and received the following -

```
C:\Program Files (x86)\Microsoft Visual Studio 14.0>peverify C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe

Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

[IL]: Error: [C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe : Prime.Observation+subscribe@38-1[b,a,w,o]::Invoke][offset 0x000000B4] Unable to resolve token.
 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

[IL]: Error: [C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe : Prime.Observation+subscribe@102-2[b,a,w,o]::Invoke][offset 0x0000008B] Unable to resolve token.
10 Error(s) Verifying C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe
```

Obviously, this is an issue with the code being generated by the F# compiler with --optimize off.
- Operating system - Windows 7
- Branch - N/A
- .NET Runtime, CoreCLR or Mono Version - .NET 4.6
- Editing Tools (e.g. Visual Studio Version) - Visual Studio 2015 Community
- Indications of severity - Rather severe - inhibits essential debugging tools, and seems to be a problem encountered by many users.
#### Contact Information

bryanedds@gmail.com

**Comment by bryanedds** [comment]
I ran the generated executable through peverify and received the following -

```C:\Program Files (x86)\Microsoft Visual Studio 14.0>peverify C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe

Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

[IL]: Error: [C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe : Prime.Observation+subscribe@38-1[b,a,w,o]::Invoke][offset 0x000000B4] Unable to resolve token.
 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

 [HRESULT 0x8007000B] - An attempt was made to load a program with an incorrect format.

[IL]: Error: [C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe : Prime.Observation+subscribe@102-2[b,a,w,o]::Invoke][offset 0x0000008B] Unable to resolve token.
10 Error(s) Verifying C:\Development\Prime\Prime\Prime\Prime\bin\Debug\Prime.exe```

Obviously, this is an issue with the code being generated by the F# compiler with --optimization off.

**Comment by bryanedds** [comment]
Thanks for marking this! It's been extremely problematic not being to effectively debug my Prime F# code library deployment because of this issue.

I wouldn't be surprised if one or two of the libraries other unit tests caused other code generation problems, so it might be worth running those once the initial issue is resolved.

If there are any questions, please contact me here or at the e-mail address above.

Thank you very much for working on this! I am available to help in any way I can!

**Comment by dsyme** [comment]
@bryanedds The best thing you could do would be to trim the example down to a minimal repro.  Just keep splitting it in two until a small program is left.  Then update the repro steps in this issue please.

**Comment by bryanedds** [comment]
I've attempted that several times, but can't seem to get that to work.

I'll look at trying that again, but please don't let my lack of progress hold up any investigation.

**Comment by dsyme** [comment]
@bryanedds Well, it's the first thing we would have to do when investigating the bug :)  

Do you mean there's really no code reduction you can do, i.e. any change means the bug goes away?

**Comment by bryanedds** [comment]
For what I've tried, yes.

Still, I may have been frustrated at the time, and therefore sloppy in the approaches I took.

I wish I had time to jump on another attempt right now, but I'm swamped with other stuff for the next few weeks.

**Comment by AviAvni** [comment]
@dsyme @bryanedds I Investigate it a little and if I changed this line https://github.com/bryanedds/Prime/blob/master/Prime/Prime/Prime/Observation.fs#L67
and 
https://github.com/bryanedds/Prime/blob/master/Prime/Prime/Prime/Observation.fs#L67
to
...<Either<'a, 'b>, Participant, 'd>...
the the code run without this exception
in ILSpy it's look like the type subscription@59-2$contract and there is more like it
when the generic type argument name is 'w something with the IL code is bad
and when I changed it to 'd it's ok

**Comment by bryanedds** [comment]
Ah, amazing guess work!

I took your idea and generalized it a bit by changing 'w to _ in lines 67, 82, 116, and 121, and now Prime works not only in all the unit tests with optimizations on, but in all the demo games that use it!

@dsyme - This might also explain why I couldn't reproduce it - I didn't bother trying to get type deduction where the first generic type is augmented with another type, and that seems to be what's causing the bad code to be generated. Since this is a frozen repository whose only purpose is to provide this error reproduction (the real Prime code now resides at https://github.com/bryanedds/Nu/tree/master/Prime (with the applied fix change list at https://github.com/bryanedds/Nu/commit/409427f714255642f431085926d1ce7f8ee2c23c), I'll leave it up until a direct fix is found in the F# compiler.

Thank you @AviAvni!

**Comment by bryanedds** [comment]
So, the **workaround** that I would suggest for this bug is to try loosening up the one or more of the type arguments with the _ wildcard at and around where the exception appears. If done correctly, it should still infer a workable secondary, if superfluous, parameter type.

**Comment by AviAvni** [comment]
@dsyme @bryanedds 
The bad IL with 'w 
```
.class abstract auto autochar serializable nested assembly beforefieldinit specialname 'subscription@59-2$contract'<a,(class Prime.EventWorld`1<!!0>) w>
       extends [mscorlib]System.Object
{
  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 06 00 00 00 00 00 ) 
} // end of class 'subscription@59-2$contract'
```
The good IL with 'd 
```
.class abstract auto autochar serializable nested assembly beforefieldinit specialname 'subscription@59-2$contract'<a>
       extends [mscorlib]System.Object
{
  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 06 00 00 00 00 00 ) 
} // end of class 'subscription@59-2$contract'
```

I think the problem is something in when generating the contract type
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/IlxGen.fs#L2669
I'm tring to debug it

**Comment by AviAvni** [comment]
@dsyme @bryanedds  After I build the compiler I try to debug it but visual studio cant load the pdbs even when i try to load them manualy

**Comment by AviAvni** [comment]
@dsyme  @bryanedds still debugging this
what I found is when getting to https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/IlxGen.fs#L3856
then with 'w the cloContractFreeTyvarSet have only c as free variable and with 'd c and d is free variables
it look like what create the Expr.TyLambda create the bug because the result from tyOfExpr in the line of the IlxGet is difference
hope i'll find it tomorrow

**Comment by AviAvni** [comment]
@dsyme  @bryanedds still debugging this
what i found is in typechecker when creating the  Expr.TyLambda  in the code
```
// Now bind the r.h.s. to the l.h.s. 
let rhsExpr = mkTypeLambda m generalizedTypars (rhsExpr,tauTy)
```
github don't show this line the code is more then 10k lines

when in the source code the type argument name is 'w the generalizedTypars not contains it
becuse of this line https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/TypeChecker.fs#L2106
I think I need to understand what is freeInEnv

**Comment by bryanedds** [comment]
Looks like you're making good progress! Thanks again!

**Comment by bryanedds** [comment]
BTW - just letting everyone involved know that the repository has moved to - https://github.com/bryanedds/Nu/tree/master/Prime . The old repository existed only for legacy reasons, and had to be deleted to prevent further confusion. The reported problem of bad code generation remains, although the suggested workaround was applied in this new repository, so to recreate the issue in this repository, the code has to be rolled back to pre-workaround.

**Comment by dsyme** [comment]
Can we have a minimized repro for this? Thanks

**Comment by dsyme** [comment]
I'll close this - @bryanedds please reopen if you can provide us with a no minimized repro provided

**Comment by dsyme** [comment]
(

**Comment by symbolsoft** [comment]
I suspect this might match another existing bug (I hope anyways!), but I don't see a great deal of similarity with #31.

I wonder if we could instead keep this open until I try the fix for #31 to see if that resolves the issue in my repo. If you close this, will doing so effectively hide the workaround from people who hit the same issue, no?

Cheers!
